### PR TITLE
Add support for DPoP

### LINKED ISSUE #188: Support DPoP
[RFC9449](https://www.rfc-editor.org/rfc/rfc9449.html#name-concept) defines  `DPoP JWT` as a way to:
 - Ensure that an issued `access_token` is used by the client (wallet) that requested it, and
 - Constraint the usage of the `access_token`

This DPoP (demonstration of Proof of Possession) must not be confused with Proof Of Possession that may be used within the credential request.

The basic requirement is:

**Token Request**
- Lib can place a request against the Token Endpoint (to obtain access_token) adding as an HTTP header a specially prepared DPoP JWT. `htm` and `htu` should point to token endpoint method and URL (no fragment, no query)
- In case OAUTH2 server supports DPoP, it will reply with
   -  `token_type` equal to `DPoP`, otherwise
   -    `bearer`
- Library should store this information (whether DPoP was supported)   

**Placing a call to Issuer Endpoints**
For accessing one of the endpoints of the issuer (that require `access_token`: issuance, batch, deferred, notification) , library must 
- generate a new DPoP JWT . `htm` and `htu` point to the method and endpoint used
- add the DPoP as an HTTP header
- Change the way access token is placed as a header. Instead of `Authorization bearer <access_token>` to `Authorization DPoP <access_token>'

### PR DESCRIPTION
**Pull request by babisRoutis**
The PR adds yet another optional parameter to VCI configuration, named `dPoPProofSigner: ProofSigner?`
This is merely a Nimbus Signer to be used for the purpose of creating DPoP JWT.

If this signer is not provided, the `Issuer` doesn't use DPoP. If it is provided then `Issuer` will change its behavior and send along with the token request a DPoP JWT header. 
OAUTH2 server that protected the credential issuer may or may not support DPoP bound access tokens. 
`Issuer` will adjust its behavior according to what the OAUTH2 server supports.

Other changes:
- [x] `AccessToken` that used to be a simple value class, has been implemented as a sealed hierarchy with two instances 
  - `AccessToken.Bearer` This indicates that OAUTH2 server doesn't support DPoP
  - `AccessToken.DPoP` This indicates that a DPoP JWT header & `Authorization DPoP <access_token>` has to be added to every call to a credential issuer endpoint

- [x] Recheck Serializable implementations. All states should be Serializble

Out of scope of this PR
- Binding authorization code (ordinary or PAR) to DPoP Key
- DPoP JWT using server-provided `nonce` 

Closes #188

### CODE REVIEW

**Comment by babisRoutis** [comment]
To partially test the implementation a Keycloak can be used having enabled the `dpop` feature (in preview status in 24.0.2).

With such a configuration, it was verified that Keycloak is able to recognize the DPoP JWT presented in its token endpoint and issue `access_token` & `refresh_token` bound to the public key presented with DPoP JWT (that is token response has `token_type` = `DPoP`)

**Comment by babisRoutis** [comment]
Some improvements committed:

- Token request uses DPoP JWT only if OAUTH2 server advertises support  via `dpop_signing_alg_values_supported` claim (check [here](https://datatracker.ietf.org/doc/html/rfc9449#name-authorization-server-metada))
- Calls to one of the issuers endpoints are placed using DPoP authentication scheme instead of bearer scheme. (not verified)
- Keep with the state of issuance process the `refresh_token` along with `access_token` (for future use)

Testing limitations
Keycloak v24.0.3 seems that 

- it doesn't support binding authorization code (simple or PAR) to DPoP key
- it doesn't support server-provided nonce (to be used in DPoP JWT in token request)

eudi-srv-pid-issuer

- It uses spring security which seems that supports only the bearer authentication scheme. So currently, we cannot call a credential issuer endpoint (which expects a bearer `access_token`) with a DPoP `access_token`

**Comment by ydanneg** [comment]
I think `DelegatingProofSigner` can be a part of API?

**Comment by babisRoutis** [comment]
> I think `DelegatingProofSigner` can be a part of API?

Sure, if you think that it can be useful.

Something like a factory method to ProofSigner compagnion object, perhaps 

```kotlin

        fun make(
            privateKey: JWK,
            algorithm: JWSAlgorithm,
            publicKey: BindingKey,
        ): ProofSigner {
            val signer = DefaultJWSSignerFactory().createJWSSigner(privateKey, algorithm)
            return object : ProofSigner , JWSSigner by signer {
                override fun getBindingKey(): BindingKey = publicKey
                override fun getAlgorithm(): JWSAlgorithm = algorithm
            }
        }
        
```

**Comment by babisRoutis** [comment]
An update: eudi-pid-issuer is being prepared to test this https://github.com/eu-digital-identity-wallet/eudi-srv-pid-issuer/pull/139

**Comment by babisRoutis** [comment]
FYI: eudi-pid-issuer now supports both bearer & DPoP authentication schemes. So it is up to the wallet (vci lib) to decide which scheme to use.
The docker image of the issuer (label `edge`) has been updated.

File: src/main/kotlin/eu/europa/ec/eudi/openid4vci/Issuance.kt [hunk 1]
```diff
@@ -346,6 +351,22 @@ interface ProofSigner : JWSSigner {
     fun getBindingKey(): BindingKey
 
     fun getAlgorithm(): JWSAlgorithm
+
+    companion object {
+
+        fun make(
+            privateKey: JWK,
+            publicKey: BindingKey,
+            algorithm: JWSAlgorithm,
```
[the remark this answers is not in the document]
Comment (babisRoutis) [reply src/main/kotlin/eu/europa/ec/eudi/openid4vci/Issuance.kt:360]: I think that this is being checked by Nimbus `DefaultJWSSignerFactory.createJWSSigner(JWK key, JWSAlgorithm alg)` method used by the `make`.

Others checks performed are:
* key is private
* key is marked to be used for signing
* algorithm to key compatibility

[the remark this answers is not in the document]
Comment (babisRoutis) [reply src/main/kotlin/eu/europa/ec/eudi/openid4vci/Issuance.kt:360]: Added the above check

**Pull request merged**